proposals/ytt/003-openapi-schemas/README.md

```diff
@@ -0,0 +1,85 @@
+# Generate OpenAPI Schema
+
+- Status: **Scoping** | Pre-Alpha | In Alpha | In Beta | GA | Rejected
+- Originating Issue: [ytt#103](https://github.com/vmware-tanzu/carvel-ytt/issues/103)
+
+# Problem Statement
+Configuration Authors want to be able to generate documentation for configuration inputs (for example, Data Values Schema, and Data Values) for their users. Additionally, Configuration Consumers want to be able to validate their configuration inputs by other tools (for example, IDE's and OpenAPI Schema validators). 
+
+# Proposal
+Implement a flag to create an [OpenAPI Schema](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md) yaml document generated from a [Data Values schema](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-schema/) file, and an optional Data Values file. When invoking ytt with configuration inputs and the flag `--schema-inspect-openapi`, the result is the OpenAPI formatted Schema object only. 
+In addition, the `--schema-inspect-openapi-metadata` flag may be used to include _both_ the [OpenAPI metadata](#openapi-document-with-metadata) and the [Schema object](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#schema-object).
+
+This flag will behave similar to the `--data-values-inspect` flag, which excludes any templates in the output.
+
+## Examples:
+### Scenario 1: Data Values Schema file only
+`ytt -f schema.yml --schema-inspect-openapi` will generate an OpenAPI Schema document using the values from the schema.yml file.
+
+### Scenario 2: Data Values Schema and Data Values
+`ytt -f schema.yml -f values.yml --schema-inspect-openapi` will generate an OpenAPI Schema document using the merged values of both the schema.yml and the values.yml file.
+
+Note: Using the merged result of both files allows the generated OpenAPI Schema to have a default value for array items, since Data Values Schema files do not yet support default values for arrays. Additionally, it allows the OpenAPI Schema to document the merged result of Data Values Schema and Data Values that will ultimately be used as configuration input during templating.
+
+### Scenario 3: Data Values file only
+`ytt -f values.yml --schema-inspect-openapi` will result in a warning suggesting that a Data Values Schema file be provided. 
+
+The reason for waiting to support Data Values files only is so that the OpenAPI document can have the additional information of default values and descriptions only available in Data Values Schema files. If this functionality comes with no additional work, we should reevaluate this decision.
```

Comment:
I personally have been viewing using a Data Values Overlay to provide defaults for an array type as a workaround rather than _the_ way.

If we had already implemented [`@schema/default`](https://hackmd.io/pODV3wzbT56MbQTxbQOOKQ?view#schemadefault) would it obviate the desire to include data values in this output?
If so, I wonder if we want to defer supporting non-empty default values in "schema inspection"... then, when we _do_ implement the ability to explicitly set a default in schema, it all "just works."

There's a little bit of implementation poking up here: schema is calculated in a completely separate phase _before_ data values are. If we design-in including data values in this inspection, it would need to occur long after schema is actually calculated.
Perhaps this is overly pedantic thinking, but shouldn't we consider schema as separate from data values?

Reply:
The main motivator for using the merged Data Values Schema _and_ Data Values in the output is so that a configuration _consumer_ can know what default values they need to override. For example, if there exists
```
#@data/values-schema 
maximum: 1
```
```
#@data/values 
maximum: 2
```
And we only use the Data Values Schema file and not the Data Values file, then the OpenAPI schema would be:
```
maximum:
  type: integer
  default: 1
```
In the case of kapp-controller, the entire configuration may not always be available to the configuration consumer, if they only read this OpenAPI Schema they would have the impression the value for `maximum` will be 1, and if that is the value they need, then they would not modify or override it.

However when invoking `ytt -f schema.yml -f values.yml` the value for `maximum` would actually be 2. Which would not be what they would expect.

Reply:
Thank you for explaining that. Yeah, I see how what's relevant to me as a Configuration Consumer is the _net_ default values. 👍🏻

Reply:
Update: While the point above about configuration consumers need to know what default values they need to override still holds true, the proposal has been updated to only include Data Values Schema files in the `--data-values-schema-inspect` command. 

> but shouldn't we consider schema as separate from data values?

Yes. Data Values Schema files are able to accomplish everything that a Data Values file can do (caveat - this will only be true once a bit more issues have been implemented), and for most cases configuration authors can use Data Values Schema files for everything they want to do with Data Values files. This should eliminate the need for Data Values overlays. 

By allowing use of only Data Value Schema files for now, we are able to take advantage of the additional information they provide that is helpful in an OpenAPI Doc. If the need for supporting Data Values overlays does present itself in the future, we can implement this flag: `ytt -f config/ --data-values-inspect -o oepnapi` which will support it.

> If we had already implemented @schema/default would it obviate the desire to include data values in this output?

As part of this work we will need to implement `#@schema/default in order to only use Data Values Schema files.